src/mds/Server.cc

```diff
@@ -5955,6 +5955,49 @@ int Server::parse_layout_vxattr(string name, string value, const OSDMap& osdmap,
   return 0;
 }
 
+int64_t Server::make_quota_max_byte_parseable(string val)
+{
+  string before_decimal;
+  string after_decimal;
+  int64_t dec_len;
+  int64_t q;
+  string cast_err;
+  
+  size_t dec_index = val.find_first_of(".");
+  size_t unit_index = val.find_first_not_of("0123456789-+.");
+  unit = str.substr(unit_index, val.length() - unit_index);
+  if (dec_index != std::string::npos) {
+    if ((unit_index == std::string::npos) || (unit == 'B'))  {
+    dout(10) << __func__ << ":  failed to parse quota.max_bytes: "
+    << "Invalid Byte value" << dendl;
+    return -CEPHFS_EINVAL;
+    }
+  }
+  dec_len = unit_index - dec_index - 1;
+  if ((dec_len == 0) || (dec_index == val.length() - 1)) {
+    dout(10) << __func__ << ":  failed to parse quota.max_bytes: "
+    << "Invalid Float value" << dendl;
+    return -CEPHFS_EINVAL;
+  }  
+  
+  if (dec_index != std::string::npos) {
+    before_decimal = val.substr(0, dec_index);
+    after_decimal = val.substr(dec_index + 1, val.length() - dec_index - 1);
+    before_decimal.append(after_decimal);
+    q = strict_iec_cast<int64_t>(before_decimal, &cast_err);
+    q = q / (std::pow(10, dec_len));
+  }
```

Comment:
if we take an example of `2..5GiB`:

the `decimal_index` will be `1`, the `unit_index` will be `4`. The check of `if ((dec_len == 0) || (dec_index == val.length() - 1)) {` would easily be bypassed. Then at `if (dec_index != std::string::npos) {` the `before_decimal` would be `2` and `after_decimal` would be `.5`. So the final string `before_decimal` would be `2.5` instead of `25` IIUC? In that case ` q = strict_iec_cast<int64_t>(before_decimal, &cast_err);` is bound to error out right?

Have you checked in your testing if values like `2..5GiB` or `....10GB` works?

Reply:
Thanks, this corner case was missing. PTAL

Reply:
If I take an example of `...10.5GiB` then:
`deciamal_index` = 0
`unit_index` = 3

`dec_len = unit_index - dec_index - 1;` passes the test since 3 - 0 - 1 = 2 then

`before_decimal` = empty str 
`after_decimal` = val.substr(1, 10 - 0 - 1) = ..10.5GiB

final output is `..10.5GiB` from `before_decimal.append(after_decimal);`. `q = strict_iec_cast<int64_t>(before_decimal, &cast_err);` would fail. Not only this, the code might also fail for `..10.GiB../`. You need to sanitize the string first and then figure out which part is digits and which one is unit.

I'd suggest you do some pre-checks like:
```
size_t first_num = val.find_first_of("0123456789")
if (first_num != std::string:npos) {
	val.erase(0, first_num);
}

// unit length is at max 3 MiB, GiB, TiB
size_t unit_begin = val.find_first_of("ABCDEFGHIJKLMNOPQRSTUVWXYZ");
string unit = val.substr(unit_begin, unit_begin + 3);

// then append
string final = val.append(unit);
```
the let the `strict_iec_cast` decide if the unit is correct

but th

Reply:
> If I take an example of `...10.5GiB` then: `deciamal_index` = 0 `unit_index` = 3
> 
> `dec_len = unit_index - dec_index - 1;` passes the test since 3 - 0 - 1 = 2 then
> 
> `before_decimal` = empty str `after_decimal` = val.substr(1, 10 - 0 - 1) = ..10.5GiB
> 
> final output is `..10.5GiB` from `before_decimal.append(after_decimal);`. `q = strict_iec_cast<int64_t>(before_decimal, &cast_err);` would fail. Not only this, the code might also fail for `..10.GiB../`. You need to sanitize the string first and then figure out which part is digits and which one is unit.
> 

Did you check the last code, I don't think this is the case with that

Reply:
> // unit length is at max 3 (e.g. MiB, GiB, TiB)
> size_t unit_begin = val.find_first_of("ABCDEFGHIJKLMNOPQRSTUVWXYZ");
> string unit = val.substr(unit_begin, unit_begin + 3);
> 
> // then append
> string final = val.append(unit);
> ```
> 
> then let the `strict_iec_cast` decide if the unit is correct

Checking the size is already being done by 'strict_iec_cast', I don't think explicitly resizing it to 3 makes sense

Reply:
I have refactored the code, please have a look. I think the cases you mentioned won't occur

Reply:
I *may* have stumbled upon an outdated version. I'll check in someitme.

Reply:
From my dry-run, I think the before_decimal.append(after_decimal) for a case like `...10.5GiB` is `..10.5GiB`. Are you sure you tested this case or similar cases and the function acted fine? IMO, this should get caught in
```
  dec_len = unit_index - dec_index - 1;
  if ((dec_len == 0) || (dec_index == val.length() - 1)) {
    dout(10) << __func__ << ":  failed to parse quota.max_bytes: "
    << "Invalid Float value" << dendl;
    return -CEPHFS_EINVAL;
  } 
```
i.e. return `Invalid Float value` to CLI but I think the case would dodge the above code and the error would be caught in `strict_iec_cast()`. We can sync on this once you're back.